tensorflow/lite/experimental/shlo/ops/convolution.cc

```diff
@@ -0,0 +1,235 @@
+          int64_t filter_spatials[rhs.Rank() - 2];
+          size_t depth = 1;
+          for (size_t m = rhs.Rank() - 1; m > 1; --m) {
+            filter_spatials[m - 2] = (l / depth) % rhs.shape().Dim(m);
+            depth *= rhs.shape().Dim(m);
+          }
+
+          // This will be replaced by tensor FlattenIndex function
+          int64_t lhs_dimensions[lhs.Rank()];
+          lhs_dimensions[0] = i;
+          lhs_dimensions[1] = k;
+          depth = 1;
+          size_t lhs_index = 0;
+          for (int64_t m = lhs.Rank() - 1; m >= 0; --m) {
+            if (m > 1) {
+              lhs_dimensions[m] =
+                  output_dimensions[m] * op.attributes.window_strides[m - 2] +
+                  filter_spatials[m - 2] * op.attributes.rhs_dilation[m - 2];
+            }
+            lhs_index += lhs_dimensions[m] * depth;
+            depth *= lhs.shape().Dim(m);
+          }
+
+          size_t channel_skip = k * rhs_spatial_size;
+          lhs_slice_pointer[l + channel_skip] = lhs_buffer[lhs_index];
+        }
+      }
+      for (size_t k = 0; k < rhs.shape().Dim(0); ++k) {
+        size_t batch_skip = k * rhs_tensor_size;
+        std::copy(rhs_buffer + batch_skip,
+                  rhs_buffer + batch_skip + rhs_tensor_size, rhs_slice_pointer);
+
+        absl::Span<Axis> lhs_batching_span;
+        absl::Span<Axis> rhs_batching_span;
+        absl::Span<Axis> lhs_contracting_span(op.lhs_contracting_dimensions);
+        absl::Span<Axis> rhs_contracting_span(op.rhs_contracting_dimensions);
+        SHLO_REF_RETURN_ON_ERROR(DotGeneralImpl<storage_type>(
+            op, lhs_slice, rhs_slice, lhs_batching_span, rhs_batching_span,
+            lhs_contracting_span, rhs_contracting_span, dot_general_output));
+
+        StorageT* dot_general_output_buffer =
+            dot_general_output.GetDataAs<storage_type>();
+
+        // This will be replaced by tensor FlattenIndex function
+        output_dimensions[0] = i;
+        output_dimensions[1] = output_channel + k;
+        output_depth = 1;
+        size_t output_index = 0;
+        for (int64_t m = output.Rank() - 1; m >= 0; --m) {
+          output_index += output_dimensions[m] * output_depth;
+          output_depth *= output.shape().Dim(m);
+        }
+        output_buffer[output_index] = dot_general_output_buffer[0];
+      }
+    }
+  }
+  output_channel += rhs.shape().Dim(0);
+
+  return absl::OkStatus();
+}
+
+template <DataType storage_type>
+absl::Status EvaluateImpl(ConvolutionOp& op, const Tensor& lhs,
+                          const Tensor& rhs, Tensor& output) {
+  size_t output_channel = 0;
+  SHLO_REF_RETURN_ON_ERROR(
+      ConvolutionImpl<storage_type>(op, output_channel, lhs, rhs, output));
+  return absl::OkStatus();
+}
+
+ConvolutionOp Create(const ConvolutionOp::Attributes& attributes) {
+  return {.attributes = attributes};
+}
+
+absl::Status Prepare(ConvolutionOp& op, const Tensor& lhs, const Tensor& rhs,
```

Comment:
For the future, I would generally suggest starting an Op with simple stubs for Create/Prepare/Eval. If small enough, we can also add constraint checking in the first PR. If too big, constraint checks should probably be the second PR. It is probably more useful to do constraint checks before doing actual implementation, as they can help narrow the implementation scope.